odoo/models.py

```diff
@@ -5281,46 +5294,77 @@ def _recompute_check(self, field):
         """
         return self.env.check_todo(field, self)
 
-    def _recompute_todo(self, field):
-        """ Mark ``field`` to be recomputed. """
-        self.env.add_todo(field, self)
-
-    def _recompute_done(self, field):
-        """ Mark ``field`` as recomputed. """
-        self.env.remove_todo(field, self)
+    @api.model
+    def recompute_fields(self, fields):
+        for fname in fields:
+            field = self._fields[fname]
+            while field and field.compute:
+                recs = self.env.field_todo(field)
+                if recs:
+                    field.compute_value(recs)
+                field = field.related_field
 
     @api.model
     def recompute(self):
         """ Recompute stored function fields. The fields and records to
             recompute have been determined by method :meth:`modified`.
         """
+        if not(self.env.recompute and self._context.get('recompute', True)):
+            return False
+
+        count = 0
+        done = {}
         while self.env.has_todo():
-            field, recs = self.env.get_todo()
-            # determine the fields to recompute
-            fs = self.env[field.model_name]._field_computed[field]
-            ns = [f.name for f in fs if f.store]
-            # evaluate fields, and group record ids by update
-            updates = defaultdict(set)
-            for rec in recs:
-                try:
-                    vals = {n: rec[n] for n in ns}
-                except MissingError:
-                    continue
-                vals = rec._convert_to_write(vals)
-                updates[frozendict(vals)].add(rec.id)
-            # update records in batch when possible
-            with recs.env.norecompute():
-                for vals, ids in updates.items():
-                    target = recs.browse(ids)
-                    try:
-                        target._write(dict(vals))
-                    except MissingError:
-                        # retry without missing records
-                        target.exists()._write(dict(vals))
-
-            # mark computed fields as done
-            for f in fs:
-                recs._recompute_done(f)
+            field = self.env.get_todo()
+            recs = self.env.field_todo(field)
+            field.compute_value(recs)
```

Comment:
There was a **major** optimization here: if all recs have the same value, it was performing **one** call to `_write`.